Cookbook for adding tabs (#549)
sulu/sulu-docs: cookbook/add-tabs.rst

Adding tabs to Sulu's Admin UI
==============================

> .. note:: It is recommended to read `Extend Admin UI`_ beforehand to get a better understandig of how Sulu admin classes work.

Author: OK will check.

> .. note:: It is recommended to read `Extend Admin UI`_ beforehand to get a better understandig of how Sulu admin classes work.

Author: Will try...

> This tutorial will walk you through the process of adding an extra tab.

Reviewer: ```suggestion
This tutorial will walk you through the process of adding an extra tab to the administration interface.
```

This could be useful in many different situations, like e.g. adding an extra tab for social media information to pages or contacts.

In this example we'll be adding a "Socials" tab to the page form of Sulu.

.. figure:: ../img/socials-tab.png

Create the form for this view
-----------------------------

You have to create the form that is rendered in this view.
Therefore create a form at ``config/forms/page_socials.xml``. 

> .. note:: Notice how we use slashes in the names of the properties, this returns the values in the given hierarchy.

Author: OK.

                </meta>
                <properties>
                    <property name="ext/social/twitter_title" type="text_line">
                        <meta>
                            <title lang="en">Twitter title</title>
                        </meta>
                    </property>
                    <property name="ext/social/twitter_description" type="text_line">
                        <meta>
                            <title lang="en">Twitter description</title>
                        </meta>
                    </property>
                    <property name="ext/social/twitter_image" type="single_media_selection">
                        <meta>
                            <title lang="en">Twitter image</title>
                        </meta>
                        <params>
                            <param name="types" value="image"/>
                            <param name="formats" type="collection">
                                <param name="og_image" />
                            </param>
                        </params>
                    </property>
                </properties>
            </section>
        </properties>
    </form>

> Registering your own admin class

Reviewer: ```suggestion
Register the view in your admin class
```

> --------------------------------

Author: Will fix

> We'll create a class for example ``src/Admin/PageAdminExtension`` which extends ``Sulu\Bundle\AdminBundle\Admin``.

Reviewer: ```suggestion
We'll create a class for example ``src/Admin/SocialAdmin`` which extends ``Sulu\Bundle\AdminBundle\Admin``.
```

> In the ``Admin`` we need to implement ``configureViews``. For our example we need the `webspace_manager` in combination with `security_checker` to check if the logged in user has permission to edit pages.

Author: Ah OK.

.. code-block:: bash

    $ php bin/console debug:container --tag=sulu.admin

        Service ID               Class name                                           
        sulu_contact.admin       Sulu\Bundle\ContactBundle\Admin\ContactAdmin        
        sulu_preview.admin       Sulu\Bundle\PreviewBundle\Admin\PreviewAdmin        
        sulu_custom_urls.admin   Sulu\Bundle\CustomUrlBundle\Admin\CustomUrlAdmin    
        sulu_website.admin       Sulu\Bundle\WebsiteBundle\Admin\WebsiteAdmin        
        sulu_tag.admin           Sulu\Bundle\TagBundle\Admin\TagAdmin                
        sulu_page.admin          Sulu\Bundle\PageBundle\Admin\PageAdmin              
        sulu_snippet.admin       Sulu\Bundle\SnippetBundle\Admin\SnippetAdmin        
        sulu_category.admin      Sulu\Bundle\CategoryBundle\Admin\CategoryAdmin      
        sulu_security.admin      Sulu\Bundle\SecurityBundle\Admin\SecurityAdmin      
        sulu_media.admin         Sulu\Bundle\MediaBundle\Admin\MediaAdmin            
        sulu_search.admin        Sulu\Bundle\SearchBundle\Admin\SearchAdmin          
        my_own.admin             App\Admin\PageAdminExtension  

> You should now see the tab show up but it's not working yet.

Reviewer: ```suggestion
You should now see the tab in the administration interface, but the data of the form is not saved yet.
```

> Registering your structure extension 

Reviewer: This is quite page specific. Should we use some more generic title like `Persist the data of the form`?
 @danrot

------------------------------------

> For this to work we need to build a structure extension handling the storage of additional data for pages.

Reviewer: What do you think about something like:
```
In the final step, we need to persist the data of the added tab. 
In the case of the pages, we can utilize the existing pages API by registering 
a new `StructureExtension`. 
In other cases, we would need to implement our own API endpoint for the tab 
as shown in the `Extend Admin UI`_ chapter.
```

> Adding tabs to Sulu admin interface

Reviewer 2: In order to better match other headings I would rephrase it to "Adding tabs to Sulu's Admin UI"

===================================

> .. note:: It is recommended to read `Extend admin Ui`_ beforehand to get a better understandig of how Sulu admin classes work.

Reviewer 2: The link is not working, and could you please change the wording to "Extend Admin UI", to match the case int he other page?

Author: Not sure why the link is not working. Can you have a look?

> 

Reviewer 2: Only one new line

> This tutorial will walk you through the process of adding extra tabs to the user interface of Sulu.

Reviewer 2: "of adding an extra tab" (you are only adding one tab in this entry, right?)

> This could be useful in a few situations, a few examples below:

Reviewer 2: The examples are almost the same... Could we maybe write something like this without a listing?

"This could be useful in many different situations, like e.g. adding an extra tab for social media information to pages or contacts."

- Saving extra data for pages
- Saving extra data in webspaces
- Saving extra data for users

> In this example we'll be adding a "Socials" tab to the PageAdmin of Sulu.

Reviewer 2: "to the page form of Sulu" - Otherwise it is not really clear, the Admin class can add to a lot of things.

> .. figure:: ../img/socials-tab.png

Reviewer 2: Can you make a screen shot of the full browser window? Otherwise it looks a little bit weird, because the toolbar is only partly visible. I think you should use a tool like "FireShot" and make the window a bit smaller.

Create the form for this view
-----------------------------

> You have to create the form that is rendered in this view. For this we create an XML file like your used to.

Reviewer 2: Add a new line between headings and texts.

> You have to create the form that is rendered in this view. For this we create an XML file like your used to.

Reviewer 2: Would omit the second sentence.

> Create your form: ``config/forms/page_socials.xml``. Notice how we use slashes in the names of the properties, this will be of later use.

Reviewer 2: "Therefore create a form at ``config/forms/page_socials.xml``:"

The notice could be an actual note using the `.. note::` syntax. Also explain what it is doing (returning the values in the given hierarchy) instead of just saying that it will be of use later.

.. code-block:: xml

    <?xml version="1.0" ?>
    <form xmlns="http://schemas.sulu.io/template/template"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/form-1.0.xsd"
    >
        <key>page_socials</key>

        <properties>
            <section name="twitter">
                <meta>

>                     <title>Twitter</title>

Reviewer 2: This should cause warnings in the UI, because if you don't pass a language to the `title` tag it will be tried to be used as a translation key, and this translation key does not exist. We could work around that by writing it the following way instead:

```xml
<title lang="en">Twitter</title>
```

                </meta>
                <properties>
                    <property name="ext/social/twitter_title" type="text_line">
                        <meta>

>                             <title>Twitter title</title>

Reviewer 2: Same as above

                        </meta>
                    </property>
                    <property name="ext/social/twitter_description" type="text_line">
                        <meta>

>                             <title>Twitter description</title>

Reviewer 2: Same as above

                        </meta>
                    </property>
                    <property name="ext/social/twitter_image" type="single_media_selection">
                        <meta>

>                             <title>Twitter image</title>

Reviewer 2: Same as above

                        </meta>
                        <params>
                            <param name="types" value="image"/>
                            <param name="formats" type="collection">
                                <param name="og_image" />
                            </param>
                        </params>
                    </property>
                </properties>
            </section>
        </properties>
    </form>

> 

Reviewer 2: Only one new line

Registering your own admin class
--------------------------------

We'll create a class for example ``src/Admin/PageAdminExtension`` which extends ``Sulu\Bundle\AdminBundle\Admin``.

> In the admin we need to implement configureViews. Also for our example we use the `webspace_manager` in combination with `security_checker` to check if the logged in user has permission to edit pages.

Reviewer 2: Put code reference in two backticks (talking about `configureViews` here)

> In the admin we need to implement configureViews. Also for our example we use the `webspace_manager` in combination with `security_checker` to check if the logged in user has permission to edit pages.

Reviewer 2: And "In the `Admin` class".

> In the admin we need to implement configureViews. Also for our example we use the `webspace_manager` in combination with `security_checker` to check if the logged in user has permission to edit pages.

Reviewer 2: "For our example we need the ..."

> Also important, notice how we set the parent to be `PageAdmin::EDIT_FORM_VIEW`. This adds our view to the `PageAdmin::EDIT_FORM_VIEW` view.

Reviewer 2: "Notice the `setParent` call adding the view as a child to `PageAdmin::EDIT_FORM_VIEW`. This will result in a new tab in the edit form for pages.

> The setFormKey is a reference to the <key> in the created form from the above section.

Reviewer 2: "`setFormKey` takes a string reference, which should be the same as the `key` tag in the above form XML."

.. code-block:: php

>     class PageAdminExtension extends Admin

Reviewer 2: I think I would call this `SocialAdmin` or something like this... But let me talk to my colleagues about that first.

Reviewer: `SocialAdmin` sounds good to me 🙂

        }

        public function configureViews(ViewCollection $viewCollection): void
        {
            $formToolbarActionsWithoutType = [
                new ToolbarAction('sulu_admin.save_with_publishing'),
            ];

            $routerAttributesToFormRequest = ['parentId', 'webspace'];
            $routerAttributesToFormMetdata = ['webspace'];

            $previewCondition = 'nodeType == 1';

            if ($this->hasSomeWebspacePermission()) {
                $viewCollection->add(
                    $this->viewBuilderFactory
                        ->createPreviewFormViewBuilder('sulu_page.page_edit_form.socials', '/socials')
                        ->disablePreviewWebspaceChooser()
                        ->setResourceKey('pages')
                        ->setFormKey('page_socials')
                        ->setTabTitle('Socials')
                        ->setTabPriority(511)
                        ->addToolbarActions($formToolbarActionsWithoutType)
                        ->addRouterAttributesToFormRequest($routerAttributesToFormRequest)
                        ->setPreviewCondition($previewCondition)
                        ->setTitleVisible(true)

>                         ->setTabOrder(1025)

Reviewer 2: Actually we tend to use binary numbers here, so that you could add more stuff in between later. So in that case it would be `2048`.

                        ->setParent(PageAdmin::EDIT_FORM_VIEW)
                );
            }
        }

        private function hasSomeWebspacePermission(): bool
        {
            foreach ($this->webspaceManager->getWebspaceCollection()->getWebspaces() as $webspace) {
                $hasWebspacePermission = $this->securityChecker->hasPermission(
                    PageAdmin::SECURITY_CONTEXT_PREFIX . $webspace->getKey(),
                    PermissionTypes::EDIT
                );

                if ($hasWebspacePermission) {
                    return true;
                }
            }

            return false;
        }
    }

We can register this class as a service and give it a tag of ``sulu.admin``, then it wil be picked up by Sulu.

.. code-block:: yaml

>     my_own.admin:

Reviewer 2: I would call that service `app.social_admin`

        class: App\Admin\PageAdminExtension
        arguments:
            - '@Sulu\Bundle\AdminBundle\Admin\View\ViewBuilderFactoryInterface'
            - '@sulu_core.webspace.webspace_manager'
            - '@sulu_security.security_checker'
        tags:
            - { name: 'sulu.admin'}
            - { name: 'sulu.context', context: 'admin' }

> When you debug the container right now your should see your own admin show up.

Reviewer 2: "you should see your own `Admin` class show up."

.. code-block:: bash

    $ php bin/console debug:container --tag=sulu.admin

        Service ID               Class name                                           
        sulu_contact.admin       Sulu\Bundle\ContactBundle\Admin\ContactAdmin        
        sulu_preview.admin       Sulu\Bundle\PreviewBundle\Admin\PreviewAdmin        
        sulu_custom_urls.admin   Sulu\Bundle\CustomUrlBundle\Admin\CustomUrlAdmin    
        sulu_website.admin       Sulu\Bundle\WebsiteBundle\Admin\WebsiteAdmin        
        sulu_tag.admin           Sulu\Bundle\TagBundle\Admin\TagAdmin                
        sulu_page.admin          Sulu\Bundle\PageBundle\Admin\PageAdmin              
        sulu_snippet.admin       Sulu\Bundle\SnippetBundle\Admin\SnippetAdmin        
        sulu_category.admin      Sulu\Bundle\CategoryBundle\Admin\CategoryAdmin      
        sulu_security.admin      Sulu\Bundle\SecurityBundle\Admin\SecurityAdmin      
        sulu_media.admin         Sulu\Bundle\MediaBundle\Admin\MediaAdmin            
        sulu_search.admin        Sulu\Bundle\SearchBundle\Admin\SearchAdmin          
        my_own.admin             App\Admin\PageAdminExtension  

> You should now see the tab show up but it's not working just now.

Reviewer 2: "not working yet"

> 

Reviewer 2: Only one new line

Registering your structure extension 
------------------------------------

> For this all to work we need to make a structure extension which handles the saving and loading of the data.

Reviewer 2: "For this to work we need to build a structure extension handling the storage of additional data for pages."

.. code-block :: php
    
    <?php

    /**
    * extends structure with social content.

>     */

Reviewer 2: Remove that comment, as it does not really add any value.

                'twitter_image' => $twitterImage,
            ];
        }

        public function export($properties, $format = null)
        {
            $data = [];
            foreach ($properties as $key => $property) {
                $value = $property;
                if (\is_bool($value)) {
                    $value = (int) $value;
                }

                $data[$key] = [
                    'name' => $key,
                    'value' => $value,
                    'type' => '',
                ];
            }

            return $data;
        }

        public function import(NodeInterface $node, $data, $webspaceKey, $languageCode, $format)
        {
            $this->setLanguageCode($languageCode, 'i18n', null);

            $this->save($node, $data, $webspaceKey, $languageCode);
        }

        public function getImportPropertyNames()
        {
            return $this->properties;
        }

        private function encodeImages(array $data)
        {
            if ($data['twitter_image']) {
                $data['twitter_image'] = json_encode($data['twitter_image']);
            }

            return $data;
        }
    }

> Also this should be registered as a service with the tag `sulu.structure.extension`.

Reviewer 2: "This service needs to be registered as a service with the tag `sulu.structure.extension`."

Author: "This class needs to be registered as a service with the tag sulu.structure.extension." 
Right?

.. code-block :: yaml

>     my_own.extension.social:

Reviewer 2: `app.social_structure_extension`

        class: App\Structure\SocialStructureExtension
        tags: { name: 'sulu.structure.extension' }

> Because we used slashes in the property names in the form. Our data is automaticly mapped to the extension data given in the view. 

Reviewer 2: That could probably be added in the note I've mentioned above, then we can remove it down here. Also, it does still not really explain what the slashes are doing.

Author: removed here

> .. figure:: ../img/socials-tab-dump.png

Reviewer 2: Wouldn't add that as a screenshot... Maybe we can add a table showing how to access which value. The thing is that the screenshot is not searchable.

Author: I removed the screenshot. Not sure what you mean by the table.
